modules/internal/LocaleTree.chpl

```diff
@@ -27,25 +27,31 @@ module LocaleTree {
   use ChapelLocale; // For declaration of rootLocale.
 
   record chpl_localeTreeRecord {
-    var left, right: locale?;
+    var left, right: locale;
+
+    proc init() {
+    }
   }
 
   pragma "locale private" var chpl_localeTree: chpl_localeTreeRecord;
 
   proc chpl_initLocaleTree() {
+    extern proc printf(s, a1, a2, a3);
     for i in LocaleSpace {
-      var left: locale? = nil;
-      var right: locale? = nil;
+      var left: unmanaged BaseLocale? = nil;
+      var right: unmanaged BaseLocale? = nil;
       var child = (i+1)*2-1;    // Assumes that indices are dense.
       if child < numLocales {
-        left = rootLocale!.getChild(child);
+        left = rootLocale.getChild(child)._instance;
+        /*printf(c"Here: %lld Left locale : %d Left instance locale : %d\n",*/
+               /*here.id, left.locale.id, left._instance.locale.id);*/
         child += 1;
         if child < numLocales then
-          right = rootLocale!.getChild(child);
+          right = rootLocale.getChild(child)._instance;
       }
-      on rootLocale!.getChild(i) {
-        chpl_localeTree.left = left;
-        chpl_localeTree.right = right;
+      on rootLocale.getChild(i) {
+        chpl_localeTree.left._instance = left;
+        chpl_localeTree.right._instance = right;
```

Comment:
It feels a little nerve-wracking to me to be poking into the `locale` type's `._instance` field here (because it breaks the abstraction, seemingly unnecessarily).  Would it work to just store the local `left` and `right` variables as `locale` values and then just use a straight assignment here instead?  (where I'm imagining that locale assignment copies `._instance` fields over?)

Alternatively (and not quite as attractive), could the localetree just store borrows of BaseLocale classes as its left/right children?

Reply:
This is for optimization (premature, maybe) purposes.

I couldn't come up with an idiom that doesn't thwart fast on optimization. Does it impact the actual time? I don't know, yet. It maybe at scale. We go from 4 regular, 4 nb, 4 fast on to 8 regular and 4 nb on. for `test/modules/sungeun/init/printInitCommCounts` that runs with 5 locales.

I have tried to store the left and right in a ref before entering the on statement, and `always RVF` ing the locale tree record with no luck.

Reply:
I actually think this tree should be ripped out, buried, and then rewritten from scratch (or replaced with something simpler), so I don't have a problem if it limps along in a slightly ugly form for a bit longer...

Reply:
(In case that was at all unclear, by "this tree" I didn't mean "your implementation of it", but the longstanding one...)

Reply:
https://github.com/Cray/chapel-private/issues/871 captures this.